packages/front-end/components/MetricAnalysis/HistogramGraph.tsx

```diff
@@ -0,0 +1,260 @@
+import React, { FC, useEffect, useState } from "react";
+import { ParentSizeModern } from "@visx/responsive";
+import { Group } from "@visx/group";
+import { scaleLinear } from "@visx/scale";
+import { AxisBottom, AxisLeft } from "@visx/axis";
+import { Bar } from "@visx/shape";
+import {
+  TooltipWithBounds,
+  useTooltip,
+  useTooltipInPortal,
+} from "@visx/tooltip";
+import { ScaleLinear } from "d3-scale";
+import styles from "@/components/Metrics/DateGraph.module.scss";
+
+interface Datapoint {
+  start: number;
+  end: number;
+  units: number;
+}
+
+type TooltipData = { x: number; y: number; d: Datapoint };
+
+interface HistogramGraphProps {
+  data: Datapoint[];
+  formatter: (value: number, options?: Intl.NumberFormatOptions) => string;
+  height?: number;
+  margin?: [number, number, number, number];
+}
+
+function getTooltipDataFromDatapoint(
+  datapoint: Datapoint,
+  data: Datapoint[],
+  innerWidth: number,
+  yScale: ScaleLinear<unknown, unknown, never>
+) {
+  const index = data.indexOf(datapoint);
+  if (index === -1) {
+    return null;
+  }
+  const x =
+    (data.length > 0 ? (index + 0.5) / (data.length + 1) : 0) * innerWidth;
+  const y = (yScale(datapoint.units ?? 0) ?? 0) as number;
+  return { x, y, d: datapoint };
+}
+
+function getTooltipContents(
+  d: TooltipData,
+  formatter: (value: number, options?: Intl.NumberFormatOptions) => string
+) {
+  return (
+    <>
+      <div className={`mb-2 ${styles.val}`}>n: {d.d.units}</div>
+      <div className="small">
+        <span className="d-inline-block" style={{ width: 40 }}>
+          start:
+        </span>{" "}
+        {formatter(d.d.start)}
+      </div>
+      <div className="small">
+        <span className="d-inline-block" style={{ width: 40 }}>
+          end:
+        </span>{" "}
+        {formatter(d.d.end)}
+      </div>
+    </>
+  );
+}
+
+const HistogramGraph: FC<HistogramGraphProps> = ({
+  data,
+  formatter,
+  height = 220,
+  margin = [15, 15, 30, 80],
+}: HistogramGraphProps) => {
+  const { containerRef, containerBounds } = useTooltipInPortal({
+    scroll: true,
+    detectBounds: true,
+  });
+
+  const [marginTop, marginRight, marginBottom, marginLeft] = margin;
+
+  const width = (containerBounds?.width || 0) + marginRight + marginLeft;
+  const yMax = height - marginTop - marginBottom;
+  const xMax = containerBounds?.width || 0;
+  const graphHeight = yMax;
+  const binWidth = xMax / data.length;
+
+  const xScale = scaleLinear({
```

Comment:
do we care about setting `undefined()` here for any unexpeted behavior from data?

Reply:
not sure, but these should be memoized (unrelated). will fix